src/eigsolve/lanczos.jl

```diff
@@ -150,3 +150,143 @@ function eigsolve(A, x₀, howmany::Int, which::Selector, alg::Lanczos;
+    numiter = 1
+
+    converged = 0
+    local normresiduals, D, U
+
+    while true
+        K = length(fact)
+        β = normres(fact)
+
+        if β < tol && K < howmany && verbosity >= WARN_LEVEL
+            msg = "Invariant subspace of dimension $(K) (up to requested tolerance `tol = $tol`), "
+            msg *= "which is smaller than the number of requested eigenvalues (i.e. `howmany == $howmany`)."
+            @warn msg
+        end
+        # BlockLanczos can access the case of K = 1 and doesn't need extra processing
+        if K >= krylovdim || β <= tol || (alg.eager && K >= howmany)
+            # Compute eigenvalues
+            # Note: Fast eigen solver for block tridiagonal matrices is not implemented yet.
+            BTD = view(fact.T, 1:K, 1:K)
+            D, U = eigen(Hermitian(BTD))
+            by, rev = eigsort(which)
+            p = sortperm(D; by=by, rev=rev)
+            D, U = permuteeig!(D, U, p)
+
+            # Detect convergence by computing the residuals
+            bs_r = fact.r_size   # The block size of the residual (decreases as the iteration goes)
+            r = residual(fact)
+            UU = U[(end - bs_r + 1):end, :]  # The last bs_r rows of U, used to compute the residuals
+            normresiduals = let R = block_inner(r, r)
+                map(u -> sqrt(real(dot(u, R, u))), cols(UU))
+            end
+            converged = 0
+            while converged < K && normresiduals[converged + 1] <= tol
+                converged += 1
+            end
+            if converged >= howmany || β <= tol  # Successfully find enough eigenvalues.
+                break
+            elseif verbosity >= EACHITERATION_LEVEL
+                @info "BlockLanczos eigsolve in iteration $numiter, Krylov dimension = $K: $converged values converged, normres = $(normres2string(normresiduals[1:howmany]))"
+            end
+        end
+
+        if K < krylovdim
+            expand!(iter, fact; verbosity=verbosity)
+            numops += 1
+        else # Shrink and restart following the shrinking method of `Lanczos`.
+            numiter >= maxiter && break
+            keep = max(floor(Int, (3 * krylovdim + 2 * converged) / (5 * bs)), 1) * bs
+            H = zeros(S, keep + bs, keep)
+            # The last bs rows of U contribute to calculate errors of Ritz values.
+            @inbounds for j in 1:keep
+                H[j, j] = D[j]
+                H[(keep + 1):end, j] = U[(K - bs + 1):K, j]
+            end
+            # Turn diagonal matrix D into a block tridiagonal matrix, and make sure 
+            # The residual of krylov subspace keeps the form of [0,..,0,R]
+            @inbounds for j in keep:-1:1
+                h, ν = householder(H, j + bs, 1:j, j)
+                H[j + bs, j] = ν
+                H[j + bs, 1:(j - 1)] .= zero(eltype(H))
+                lmul!(h, H)
+                rmul!(view(H, 1:(j + bs - 1), :), h')
+                rmul!(U, h')
+            end
```

Comment:
After this step, `H[1:keep, 1:keep]` should still be Hermitian, but I guess that property can be lost at the numerical precision level resulting from the sequential left and right multiplication with Householder transformations. It probably doesn't matter, since eventually after growing H is passed to a Hermitian eigensolver which will anyway assume that it is exactly Hermitian. Nonetheless, maybe it is nice to restore exact hermiticity when it is copied back in to `BTD` (see next comment):